README.md

# longmixr

<!-- badges: start -->
[![R-CMD-check](https://github.com/cellmapslab/longmixr/workflows/R-CMD-check/badge.svg)](https://github.com/cellmapslab/longmixr/actions)
[![CRAN status](https://www.r-pkg.org/badges/version/longmixr)](https://CRAN.R-project.org/package=longmixr)
<!-- badges: end -->

The goal of longmixr is to provide consensus clustering for longitudinal clustering
performed with [`flexmix`](https://cran.r-project.org/package=flexmix). It uses the approach from [`ConsensusClusterPlus`](https://bioconductor.org/packages/release/bioc/html/ConsensusClusterPlus.html) but
replaces the clustering of the longitudinal data with a `flexmix` model.

## Installation
You can install longmixr from CRAN with:

``` r
install.packages("longmixr")
```

You can install the latest version from github with:

``` r
remotes::install_github("cellmapslab/longmixr")
```

If you want to render the vignette, use:

``` r
remotes::install_github("cellmapslab/longmixr", build_vignettes = TRUE, dependencies = TRUE)
```

Please note that for the vignette a lot more dependencies are installed.

## Basic usage
You need a dataset with a column that identifies the subject, a column that
denotes the time point of the measurement and variables that you want to model.

``` r
set.seed(5)
test_data <- data.frame(
  patient_id = rep(1:10, each = 4),
  visit = rep(1:4, 10),
  var_1 = c(rnorm(20, -1), rnorm(20, 3)) + rep(seq(from = 0, to = 1.5, length.out = 4), 10),
  var_2 = c(rnorm(20, 0.5, 1.5), rnorm(20, -2, 0.3)) + rep(seq(from = 1.5, to = 0, length.out = 4), 10)
)
```

In the following approach, the variables `var_1` and `var_2` each are modeled as
dependent on a smooth function of time, taking the multiple measurements for each
subject into account. The assumption is that `var_1` and `var_2` represent a
multivariate outcome. The modeling is specified in the `flexmix` drivers
and the `flexmix_formula`:

``` r
model_list <- list(flexmix::FLXMRmgcv(as.formula("var_1 ~ .")),
                   flexmix::FLXMRmgcv(as.formula("var_2 ~ .")))
clustering <- longitudinal_consensus_cluster(
  data = test_data,
  id_column = "patient_id",
  max_k = 2,
  reps = 3,
  model_list = model_list,
  flexmix_formula = as.formula("~s(visit, k = 4) | patient_id"))
```

The results of the clustering can be assessed via several plots. For every
specified number of clusters, the consensus matrix and the resulting hierarchical
clustering on this matrix is shown. Additionally, the consensus CDF and the delta
Area plots give a measure which number of cluster is optimal. The tracking plot
gives an overview how the observations are distributed across the different clusters
for different numbers of specified clusters. The item (subject) consensus
plot shows the average consensus of each subject with all other subjects that
belong to one cluster. The cluster consensus plot depicts the average consensus
between all members of each cluster.

The above mentioned plots are generated when calling the `plot` function:

``` r
plot(clustering)
```

You can also select which plots you want to generate:

``` r
plot(clustering, which_plots = "consensusmatrix_2")
```

![consensus matrix plot showing two clusters cleanly separated, one bigger than the other](man/figures/example_consensusmatrix_2.png)

### Detailed explanation
For a detailed explanation how you can use `longmixr` to analyze your
longitudinal data, check out the
[Example clustering analysis vignette](https://cellmapslab.github.io/longmixr/articles/analysis_workflow.html).

## Cross-sectional clustering
Additionally, this package provides a wrapper function around the
`ConsensusClusterPlus` function to work with mixed continuous and categorical
data (by using the Gower distance):

``` r
dc <- mtcars
# scale continuous variables
dc <- sapply(mtcars[, 1:7], scale)
# code factor variables
dc <- cbind(as.data.frame(dc),
            vs = as.factor(mtcars$vs),
            am = as.factor(mtcars$am),
            gear = as.factor(mtcars$gear),
            carb = as.factor(mtcars$carb))
cc <- crosssectional_consensus_cluster(
  data = dc,
  reps = 10,
  seed = 1
)
```

## Copyright
The package is based on the code of [`ConsensusClusterPlus`](https://bioconductor.org/packages/release/bioc/html/ConsensusClusterPlus.html)
(version 1.52.0). For this code the copyright holders are Matt Wilkerson and
Peter Waltman. For all subsequent changes the copyright holder is the Max Planck
Institute of Psychiatry. The code is licensed under GPL v2.


NEWS.md
# longmixr 1.1.0

* include the argument `which_plots` in the `plot.lcc` function so that one can
  specify which plots should be plotted
* include the argument `n_item_consensus` in the `plot.lcc` function so that one
  can specify how many item consensus plots should be plotted in one plot

# longmixr 1.0.0

This is the first version of the package (new release).


tests/testthat.R
library(testthat)
library(longmixr)

test_check("longmixr")


tests/testthat/test_ccc.R
dc <- mtcars
# scale continuous variables
dc <- sapply(mtcars[, 1:7], scale)
# code factor variables
dc <- cbind(as.data.frame(dc),
            vs = as.factor(mtcars$vs),
            am = as.factor(mtcars$am),
            gear = as.factor(mtcars$gear),
            carb = as.factor(mtcars$carb))

# as here I basically only use functions from other packages, I only test if
# the d argument is set correctly

# set up a function to run the function that calls ConsensusClusterPlus
# to manually turn off any device or changed plotting codes
# set pdf(NULL) because otherwise the test leaves a Rplots.pdf file
run_ccc <- function(...) {
  pdf(NULL)
  op <- par(no.readonly = TRUE)
  on.exit(par(op))
  on.exit(dev.off(), add = TRUE)

  crosssectional_consensus_cluster(...)
}

test_that("the d argument is set correctly", {
  # this base case should work
  expect_type(run_ccc(
    data = dc,
    reps = 3,
    seed = 1
  ),
  "list")

  # no extra d argument is allowed
  expect_error(run_ccc(
    data = dc,
    d = dc,
    reps = 3,
    seed = 1
  ))
})


tests/testthat/test_lcc_plot.R
set.seed(5)
test_data <- data.frame(patient_id = rep(1:15, each = 4),
                        visit = rep(1:4, 15),
                        var_1 = c(rnorm(20, -1), rnorm(20, 3), rnorm(20, 0.5)) +
                          rep(seq(from = 0, to = 1.5, length.out = 4), 15),
                        var_2 = c(rnorm(20, 0.5, 1.5), rnorm(20, -2, 0.3),
                                  rnorm(20, 4, 0.5)) +
                          rep(seq(from = 1.5, to = 0, length.out = 4), 15))
model_list <- list(flexmix::FLXMRmgcv(as.formula("var_1 ~ .")),
                   flexmix::FLXMRmgcv(as.formula("var_2 ~ .")))

clustering <- longitudinal_consensus_cluster(
  data = test_data,
  id_column = "patient_id",
  max_k = 3,
  reps = 3,
  model_list = model_list
)

# helper function to save the plots
# because plot.lcc outputs several plots at once, I store them all in one big
# png and compare this png
save_png <- function(code, width = 800, height = 800) {
  path <- tempfile(fileext = ".png")
  png(path, width = width, height = height)
  op <- par(mfrow = c(4, 2))
  on.exit(dev.off())
  on.exit(par(op), add = TRUE)
  code

  path
}

test_that("a plot is generated in the CI", {
  skip_if_not(isTRUE(as.logical(Sys.getenv("CI"))),
              message = "simple plot generation only tested on CI")
  test_plot <- plot(clustering)
  expect_equal(test_plot, matrix(c(0.7, 1.9, 3.1, 4.3, 5.5, 6.7, 7.9),
                                 nrow = 7))
})

test_that("the plots stay the same", {
  skip_on_ci()
  # because plot.lcc changes the graphics parameters, currently only the last
  # plot is recorded. Therefore, I can only check the last plot.
  expect_snapshot_file(save_png(plot(clustering)), "plot_lcc_output.png")
})

test_that("plot.lcc only uses lcc objects", {
  expect_error(plot.lcc(1:10))
})

test_that("the color_palette argument is correct", {
  expect_error(plot(clustering, color_palette = 1:10),
               regexp = "Assertion on 'color_palette' failed")

  expect_silent(plot(clustering,
                     color_palette = c("#9999FF", "#7F7FFF", "#6666FF")))
})

test_that("the which_plots argument is correct", {
  expect_error(plot(clustering, which_plots = 3))
  expect_error(plot(clustering, which_plots = "consensusmatrix_4"),
               regexp = "which_plot must be one of all, consensusmatrix_legend, consensusmatrix, consensusmatrix_2, consensusmatrix_3, CDF, delta, cluster_tracking, item_consensus, cluster_consensus.")
  skip_on_ci()
  # because plot.lcc changes the graphics parameters, currently only the last
  # plot is recorded. Therefore, I can only check the last plot for the
  # consensus plots, even though I specified several plots.
  expect_snapshot_file(save_png(plot(clustering, which_plots = "consensusmatrix")),
                       "plot_lcc_output_consensusmatrix.png")
  expect_snapshot_file(save_png(plot(clustering, which_plots = "delta")),
                       "plot_lcc_output_delta.png")
  expect_snapshot_file(save_png(plot(clustering, which_plots = c("CDF", "cluster_tracking"))),
                       "plot_lcc_output_cdf_cluster_tracking.png")
})

test_that("the n_item_consensus argument is correct", {
  expect_error(plot(clustering, n_item_consensus = "test"))
  expect_error(plot(clustering, n_item_consensus = -1))

  skip_on_ci()
  # only the last plot (with k=3) is recorded
  expect_snapshot_file(save_png(plot(clustering, which_plots = "item_consensus",
                                     n_item_consensus = 1)),
                       "plot_lcc_output_item_consenus_1.png")
})


R/data.R
#' Fake questionnaire data
#'
#' A simulated data set containing observations of 100 individuals at four time
#' points. The data was simulated in two groups (50 individuals each) and
#' contains two questionnaires with five items each, one questionnaire with
#' five continuous variables and one additional cross-sectional continuous
#' variable. In this data set the group variable from the simulation is
#' included. You typically don't have this group variable in your data.
#'
#' @format A data frame with 400 rows and 20 variables:
#' \describe{
#'   \item{ID}{patient ID}
#'   \item{visit}{time point of the observation}
#'   \item{group}{to which simulated group the observation belongs to}
#'   \item{age_visit_1}{age of the patient at time point 1}
#'   \item{single_continuous_variable}{a cross-sectional continuous variable,
#'   i.e. there is only one unique value per individual}
#'   \item{questionnaire_A_1}{the first item of questionnaire A with categories
#'   1 to 5}
#'   \item{questionnaire_A_2}{the second item of questionnaire A with categories
#'   1 to 5}
#'   \item{questionnaire_A_3}{the third item of questionnaire A with categories
#'   1 to 5}
#'   \item{questionnaire_A_4}{the fourth item of questionnaire A with categories
#'   1 to 5}
#'   \item{questionnaire_A_5}{the fifth item of questionnaire A with categories
#'   1 to 5}
#'   \item{questionnaire_B_1}{the first item of questionnaire B with categories
#'   1 to 5}
#'   \item{questionnaire_B_2}{the second item of questionnaire B with categories
#'   1 to 5}
#'   \item{questionnaire_B_3}{the third item of questionnaire B with categories
#'   1 to 5}
#'   \item{questionnaire_B_4}{the fourth item of questionnaire B with categories
#'   1 to 5}
#'   \item{questionnaire_B_5}{the fifth item of questionnaire B with categories
#'   1 to 5}
#'   \item{questionnaire_C_1}{the first continuous variable of questionnaire C}
#'   \item{questionnaire_C_2}{the second continuous variable of questionnaire C}
#'   \item{questionnaire_C_3}{the third continuous variable of questionnaire C}
#'   \item{questionnaire_C_4}{the fourth continuous variable of questionnaire C}
#'   \item{questionnaire_C_5}{the fifth continuous variable of questionnaire C}
#' }
#' @source simulated data
"fake_questionnaire_data"


R/crosssectional_consensus_cluster.R
#' Cross-sectional clustering with categorical variables
#'
#' This function uses the \code{ConsensusClusterPlus} function from the package
#' with the same name with defaults for clustering data with categorical
#' variables. As the distance function, the Gower distance is used.
#'
#' @param data a matrix or data.frame containing variables that should be used
#' for computing the distance. This argument is passed to \code{StatMatch::gower.dist}
#' @param reps number of repetitions, same as in \code{ConsensusClusterPlus}
#' @param finalLinkage linkage method for final clustering,
#' same as in \code{ConsensusClusterPlus}same as in \code{ConsensusClusterPlus}
#' @param innerLinkage linkage method for clustering steps,
#' same as in \code{ConsensusClusterPlus}
#' @param ... other arguments passed to \code{ConsensusClusterPlus}, attention:
#' the \code{d} argument can \bold{not} be set as it is directly computed by
#' \code{crosssectional_consensus_cluster}
#'
#' @details
#' \code{data} can take all input data types that \code{\link[StatMatch]{gower.dist}}
#' can handle, i.e. \code{numeric}, \code{character}/\code{factor}, \code{ordered}
#' and \code{logical}.
#'
#' @return The output is produced by \code{ConsensusClusterPlus}
#'
#' @importFrom stats as.dist
#' @export
#'
#' @examples
#' dc <- mtcars
#' # scale continuous variables
#' dc <- sapply(mtcars[, 1:7], scale)
#' # code factor variables
#' dc <- cbind(as.data.frame(dc),
#'             vs = as.factor(mtcars$vs),
#'             am = as.factor(mtcars$am),
#'             gear = as.factor(mtcars$gear),
#'             carb = as.factor(mtcars$carb))
#' cc <- crosssectional_consensus_cluster(
#'   data = dc,
#'   reps = 10,
#'   seed = 1
#' )
crosssectional_consensus_cluster <- function(
  data,
  reps = 1000,
  finalLinkage = "ward.D2",
  innerLinkage = "ward.D2",
  ...) {

  data_as_distance <- as.dist(StatMatch::gower.dist(data))
  consensus_cluster <- ConsensusClusterPlus::ConsensusClusterPlus(
    d = data_as_distance,
    reps = reps,
    # this argument is actually not needed as it doesn't do anything when d is
    # a distance object (as it is the case here)
    distance = "StatMatch::gower.dist",
    finalLinkage = finalLinkage,
    innerLinkage = innerLinkage,
    ...)

  consensus_cluster
}


R/longitudinal_consensus_cluster.R
#' Longitudinal consensus clustering with flexmix
#'
#' This function performs longitudinal clustering with flexmix. To get robust
#' results, the data is subsampled and the clustering is performed on this
#' subsample. The results are combined in a consensus matrix and a final
#' hierarchical clustering step performed on this matrix. In this, it follows
#' the approach from the \code{ConsensusClusterPlus} package.
#'
#' @param data a \code{data.frame} with one or several observations per subject.
#' It needs to contain one column that specifies to which subject the entry (row)
#' belongs to. This ID column is specified in \code{id_column}. Otherwise, there
#' are no restrictions on the column names, as the model is specified in
#' \code{flexmix_formula}.
#' @param id_column name (character vector) of the ID column in \code{data} to
#' identify all observations of one subject
#' @param max_k maximum number of clusters, default is \code{3}
#' @param reps number of repetitions, default is \code{10}
#' @param p_item fraction of samples contained in subsampled sample, default is
#' \code{0.8}
#' @param model_list either one \code{flexmix} driver or a list of \code{flexmix}
#' drivers of class \code{FLXMR}
#' @param flexmix_formula a \code{formula} object that describes the \code{flexmix}
#' model relative to the formula in the flexmix drivers (the dot in the flexmix
#' drivers is replaced, see the example). That means that you usually only
#' specify the right-hand side of the formula here. However, this is not enforced
#' or checked to give you more flexibility over the \code{flexmix} interface
#' @param title name of the clustering; used if \code{writeTable = TRUE}
#' @param final_linkage linkage used for the last hierarchical clustering step on
#' the consensus matrix; has to be \code{average, ward.D, ward.D2, single, complete, mcquitty, median}
#' or \code{centroid}. The default is \code{average}
#' @param seed seed for reproducibility
#' @param verbose \code{boolean} if status messages should be displayed.
#' Default is \code{FALSE}
#'
#' @details
#' The data types \code{longitudinal_consensus_cluster} can handle depends on
#' how the \code{flexmix} models are set up, in principle all data types are
#' supported for which there is a \code{flexmix} driver with the desired
#' outcome variable.
#'
#' If you follow the dimension reduction approach outlined in
#' \code{vignette("Example clustering analysis", package = "longmixr")}, the
#' input data types depend on what \code{FAMD} from the \code{FactoMineR}
#' package can handle. \code{FAMD} accepts \code{numeric} variables and treats
#' all other variables as \code{factor} variables which it can handle as well.
#'
#' @return An object (list) of class \code{lcc} with length \code{maxk}.
#' The first entry \code{general_information} contains the entries:\tabular{ll}{
#'    \code{consensus_matrices} \tab a list of all consensus matrices (for all specified clusters) \cr
#'    \tab \cr
#'    \code{cluster_assignments} \tab a \code{data.frame} with an ID column named after \code{id_column} and a column for every specified number of clusters, e.g. \code{assignment_num_clus_2} \cr
#'    \tab \cr
#'    \code{call} \tab the call/all arguments how \code{longitudinal_consensus_cluster} was called
#' }
#'
#' The other entries correspond to the number of specified clusters (e.g. the
#' second entry corresponds to 2 specified clusters) and each contains a list with the
#' following entries:\tabular{ll}{
#'    \code{consensus_matrix} \tab the consensus matrix \cr
#'    \tab \cr
#'    \code{consensus_tree} \tab the result of the hierarchical clustering on the consensus matrix \cr
#'    \tab \cr
#'    \code{consensus_class} \tab the resulting class for every observation \cr
#'    \tab \cr
#'    \code{found_flexmix_clusters} \tab a vector of the actual found number of clusters by \code{flexmix} (which can deviate from the specified number)
#' }
#'
#' @importFrom stats as.formula hclust as.dist cutree
#' @importFrom utils write.csv write.table
#'
#'
#' @export
#'
#' @examples
#' set.seed(5)
#' test_data <- data.frame(patient_id = rep(1:10, each = 4),
#' visit = rep(1:4, 10),
#' var_1 = c(rnorm(20, -1), rnorm(20, 3)) +
#' rep(seq(from = 0, to = 1.5, length.out = 4), 10),
#' var_2 = c(rnorm(20, 0.5, 1.5), rnorm(20, -2, 0.3)) +
#' rep(seq(from = 1.5, to = 0, length.out = 4), 10))
#' model_list <- list(flexmix::FLXMRmgcv(as.formula("var_1 ~ .")),
#' flexmix::FLXMRmgcv(as.formula("var_2 ~ .")))
#' clustering <- longitudinal_consensus_cluster(
#' data = test_data,
#' id_column = "patient_id",
#' max_k = 2,
#' reps = 3,
#' model_list = model_list,
#' flexmix_formula = as.formula("~s(visit, k = 4) | patient_id"))
#' # not run
#' # plot(clustering)
#' # end not run
longitudinal_consensus_cluster <- function(data = NULL,
                                           id_column = NULL,
                                           max_k = 3,
                                           reps = 10,
                                           p_item = 0.8,
                                           model_list = NULL,
                                           flexmix_formula = as.formula("~s(visit, k = 4) | patient_id"),
                                           title = "untitled_consensus_cluster",
                                           final_linkage = c("average", "ward.D", "ward.D2", "single", "complete",
                                                             "mcquitty", "median", "centroid"),
                                           seed = 3794,
                                           verbose = FALSE) {

  # check variables
  checkmate::assert_count(seed, positive = TRUE)
  set.seed(seed)

  checkmate::assert_character(id_column, len = 1, any.missing = FALSE)
  checkmate::assert_data_frame(data, all.missing = FALSE, min.rows = 1, min.cols = 1)
  checkmate::assert_choice(id_column, colnames(data))
  checkmate::assert_integerish(max_k, len = 1, lower = 2)
  checkmate::assert_count(reps, positive = TRUE)
  checkmate::assert_number(p_item, lower = 1 / nrow(data), upper = 1)
  # model_list can be either a list of flexmix drivers or a single flexmix
  # driver
  checkmate::assert(checkmate::check_list(model_list, types = "FLXMR"),
                    checkmate::check_class(model_list, "FLXMR"))
  checkmate::assert_class(flexmix_formula, "formula")
  checkmate::assert_character(title, len = 1)
  final_linkage <- match.arg(final_linkage)
  checkmate::assert_logical(verbose, len = 1)

  call <- match.call()

  # perform the longitudinal clustering to create the consensus matrices
  results <- lcc_run(data = data,
                     id_column = id_column,
                     max_k = max_k,
                     reps = reps,
                     p_item = p_item,
                     model_list = model_list,
                     flexmix_formula = flexmix_formula,
                     verbose = verbose)

  consensus_matrices <- results[["consensus_matrices"]]
  flexmix_found_clusters <- results[["found_number_clusters"]]

  # generate the consensus clustering on the consensus matrices
  # (last clustering step)
  res <- vector(mode = "list", length = max_k)
  # do the final clustering for every specified number of clusters
  for (cluster_index in 2:max_k) {
    if (verbose) {
      message(paste("consensus ", cluster_index))
    }
    consensus_matrix <- consensus_matrices[[cluster_index]]
    hc_tree <- hclust(as.dist(1 - consensus_matrix), method = final_linkage)
    if (verbose) {
      message("clustered")
    }
    cut_tree_groups <- cutree(hc_tree, cluster_index)
    names(cut_tree_groups) <- sort(unique(data[, id_column]))

    res[[cluster_index]] <- list(consensus_matrix = consensus_matrix,
                                 consensus_tree = hc_tree,
                                 consensus_class = cut_tree_groups,
                                 found_flexmix_clusters = flexmix_found_clusters[[cluster_index]])
  }

  # generate a data.frame with all cluster assignments for the subjects
  assignment_table <- extract_assignment(results = res,
                                         id_column = id_column)

  res[[1]] <- list(consensus_matrices = consensus_matrices,
                   cluster_assignments = assignment_table,
                   call = call)
  names(res)[1] <- "general_information"

  # create meaningful names for the different clustering solutions
  if (length(res) > 1) {
    names(res)[-1] <- paste0("cluster_", seq(from = 2, to = length(res), by = 1))
  }

  class(res) <- c("lcc", class(res))
  return(res)
}

#' Main function to run the longitudinal clustering
#'
#' Internal function to actually perform the clustering
#'
#' @inheritParams longitudinal_consensus_cluster
#'
#' @return Returns a list with the following entries:\tabular{ll}{
#'    \code{consensus_matrices} \tab a list of all consensus matrices where the kth entry is the consensus matrix for k specified numbers of clusters. The first entry is \code{NULL} \cr
#'    \tab \cr
#'    \code{found_number_clusters} \tab a list of vectors with the actual number of clusters found by \code{flexmix} where the kth entry is for k specified numbers of clusters. The first entry is \code{NULL}
#' }
#' @noRd
lcc_run <- function(data,
                    id_column,
                    max_k,
                    reps,
                    p_item,
                    model_list,
                    flexmix_formula,
                    verbose) {
  # internal function, therefore no input checks

  connectivity_results <- vector(mode = "list", max_k)
  # get the number of unique patients
  unique_patient_ids <- sort(unique(data[, id_column]))
  n <- length(unique_patient_ids)
  # initialise the connectivity matrix
  sample_count_matrix <- matrix(0, ncol = n, nrow = n)
  # be lazy and initialise the list with 0, because the first element corresponds
  # to a cluster size of 1
  connectivity_results[[1]] = 0
  # initialise list for found number of clusters
  found_number_clusters <- vector(mode = "list", max_k)

  for (i in 1:reps) {
    if (verbose) {
      message(paste("random subsample", i))
    }
    sample_x = sample_patients(data = data,
                               p_samp = p_item,
                               id_column = id_column)

    # save the matrix how often a combination of 2 samples occurred in this
    # subsampled data set to correct that not all samples are contained in the
    # subsample
    sample_count_matrix <- connectivity_matrix(cluster_assignments =
                                                 rep(1, length(sample_x[["sample_patients"]])),
                                               current_matrix = sample_count_matrix,
                                               matrix_names = unique_patient_ids,
                                               sample_key = sample_x[["sample_patients"]])

    # cluster the subsampled data
    fitted_models <- flexmix::stepFlexmix(formula = flexmix_formula,
                                          data = sample_x[["subsample"]],
                                          k = seq(from = 2, to = max_k, by = 1),
                                          model = model_list,
                                          nrep = 1)

    for (cluster_index in seq(from = 2, to = max_k, by = 1)) {

      if (i == 1) {
        # initialise the connectivity matrix during the first time
        connectivity_results[[cluster_index]] <- matrix(0, ncol = n, nrow = n)
      }

      # extract the cluster assignments from flexmix
      if (inherits(fitted_models, "stepFlexmix")) {
        # when more than 1 cluster values are specified, the returned value is
        # a stepFlexmix object and the models can be accessed by the cluster
        # number
        assignment_all_values <- fitted_models@models[[as.character(cluster_index)]]@cluster
        # extract the actual found number of clusters (as flexmix might find
        # a cluster solution with less clusters than specified)
        actual_number_found_clusters <- fitted_models@models[[as.character(cluster_index)]]@k

      } else if (inherits(fitted_models, "flexmix") && fitted_models@k == as.integer(cluster_index)) {
        # if only one cluster value was used during fitting (i.e. only k=2)
        assignment_all_values <- fitted_models@cluster
        actual_number_found_clusters <- fitted_models@k
      } else {
        stop("Something went wrong with extracting the cluster assignments from the flexmix models")
      }

      merged_data <- data.frame(patient_id = sample_x[["subsample"]][, id_column],
                                cluster = assignment_all_values)
      # get the cluster assignment together with the corresponding ID
      # only one value for every patient (the values should be the same for every
      # repeated measurement)
      merged_data <- merged_data[!duplicated(merged_data[, "patient_id"]), ]

      # update the connectivity matrix
      connectivity_results[[cluster_index]] <- connectivity_matrix(cluster_assignments = merged_data$cluster,
                                                                   current_matrix = connectivity_results[[cluster_index]],
                                                                   matrix_names = unique_patient_ids,
                                                                   sample_key = merged_data$patient_id)
      # store the found number of clusters
      found_number_clusters[[cluster_index]] <- c(found_number_clusters[[cluster_index]],
                                                  actual_number_found_clusters)
    }
  }

  # calculate the final consensus matrices/fraction
  res <- vector(mode = "list", max_k)
  for (cluster_index in 2:max_k) {
    tmp <- triangle(connectivity_results[[cluster_index]], mode = 3)
    tmp_count <- triangle(sample_count_matrix, mode = 3)
    # tmp is a matrix of counts how often each pair of samples is in one cluster
    # temp_count is a matrix of counts how often (maximum) each pair of samples
    # could have been in a cluster
    res[[cluster_index]] <- tmp / tmp_count
    res[[cluster_index]][which(tmp_count == 0)] <- 0
  }

  list(consensus_matrices = res,
       found_number_clusters = found_number_clusters)
}

#' Try out different linkage methods
#'
#' In the final step, the consensus clustering performs a hierarchical clustering
#' step on the consensus cluster. This function tries out different linkage
#' methods and returns the corresponding clusterings. The outputs can be plotted
#' like the results from \code{\link{longitudinal_consensus_cluster}}.
#'
#' @param results clustering result of class \code{lcc}
#' @param use_methods character vector of one or several items of \code{average},
#' \code{ward.D}, \code{ward.D2}, \code{single}, \code{complete}, \code{mcquitty},
#' \code{median} or \code{centroid}
#'
#' @return a list of elements, each element of class \code{lcc}. The entries are
#' named after the used linkage method.
#'
#' @importFrom stats hclust as.dist cutree
#' @export
#'
#' @examples
#' set.seed(5)
#' test_data <- data.frame(patient_id = rep(1:10, each = 4),
#' visit = rep(1:4, 10),
#' var_1 = c(rnorm(20, -1), rnorm(20, 3)) +
#' rep(seq(from = 0, to = 1.5, length.out = 4), 10),
#' var_2 = c(rnorm(20, 0.5, 1.5), rnorm(20, -2, 0.3)) +
#' rep(seq(from = 1.5, to = 0, length.out = 4), 10))
#' model_list <- list(flexmix::FLXMRmgcv(as.formula("var_1 ~ .")),
#' flexmix::FLXMRmgcv(as.formula("var_2 ~ .")))
#' clustering <- longitudinal_consensus_cluster(
#' data = test_data,
#' id_column = "patient_id",
#' max_k = 2,
#' reps = 3,
#' model_list = model_list,
#' flexmix_formula = as.formula("~s(visit, k = 4) | patient_id"))
#'
#' clustering_linkage <- test_clustering_methods(results = clustering,
#' use_methods = c("average", "single"))
#' # not run
#' # plot(clustering_linkage[["single"]])
#' # end not run
test_clustering_methods <- function(results,
                                    use_methods = c("average", "ward.D", "ward.D2", "single", "complete",
                                                    "mcquitty", "median", "centroid")) {
  checkmate::assert_class(results, "lcc")
  checkmate::assert_character(use_methods, unique = TRUE)

  # try out all specified linkage methods
  new_results <- lapply(use_methods, function(current_method) {
    # generate the consensus clustering on the consensus matrices
    curr_res <- list()
    for (tk in seq(from = 2, to = length(results), by = 1)) {

      fm <- results[[tk]][["consensus_matrix"]]
      hc <- hclust(as.dist(1 - fm), method = current_method)
      ct <- cutree(hc, tk)
      names(ct) <- names(results[[tk]][["consensus_class"]])

      curr_res[[tk]] <- list(consensus_matrix = fm,
                             consensus_tree = hc,
                             consensus_class = ct,
                             found_flexmix_clusters = results[[tk]][["found_flexmix_clusters"]])
    }

    # add the correct general information
    original_assignments <- results[["general_information"]][["cluster_assignments"]]
    id_column <- colnames(original_assignments)[1]
    assignment_table <- extract_assignment(results = curr_res,
                                           id_column = id_column)

    # gather all consensus matrices to one list
    consensus_matrices <- list()
    for (i in seq(from = 2, to = length(results), by = 1)) {
      consensus_matrices[[i]] <- curr_res[[i]][["consensus_matrix"]]
    }
    curr_res[[1]] <- list(consensus_matrices = consensus_matrices,
                          cluster_assignments = assignment_table)
    names(curr_res)[1] <- "general_information"

    class(curr_res) <- c("lcc", class(curr_res))
    curr_res
  })

  names(new_results) <- use_methods
  new_results
}
